### ISSUE TITLE
Add support for anchor links

### ISSUE THREAD

**Issue by vigenere23** [ItalyPaleAle/svelte-spa-router#157]
Anchor links are the trailing `#` part in URL that usually corresponds to an ID on the page. The default and standard browser behavior is to automatically scroll to the element that corresponds to that id on page load. But since our app is an SPA, the `hash` starts at the first `#` symbol and thus preventing such behavior (ex: `#/#about` will try to scroll to element with id `/#about` instead of `about`).

Here is a way to achieve such a thing:

```js
const elementId = window.location.hash.split('#').pop()

    if (elementId) {
      const element = document.getElementById(elementId)
      if (element) {
        element.scrollIntoView(true) // `true` is for aligning top of element to top of screen
      }
    }
```

This would need to be inserted at the right place.

Note that currently (with regex path params implemented), this works well inside a `routeLoaded()` event handler (even with `restoreScrollState` activated) without any glitch. 

### Prerequisites

- Support for regex path params

**Comment by ItalyPaleAle** [comment]
This is an interesting issue and I'd be willing to support looking for `#` in the querystring (so a `#` after a `#`) to determine the anchor to navigate to.

i agree that this would break links that want to point to an anchor. We _could_ fix that for when people leverage the `use:link` action. There would be no way to fix it for "regular" links.

What do you think?

**Comment by vigenere23** [comment]
Yes, sounds good! But I'm wondering if the steps from `2.1` would work in the following algorithm to work around the "same-page anchor-links":

1. always keep a reference to the last hash
2. on `hashchange` event, do:
    1. if hash is an anchor link (aka there's no `/` after the `#`): 
        1. (needed?) prevent default behavior (would a `event.preventDefault()` work for this event?)
        2. change the `window.location.hash` to `lastHash + anchorLink`
        3. return (to pass to the next iteration of `hashchange` event handler). Should we return `true`, `false` or nothing?
    2. else, proceed normally but with added anchor link support:
        1. split hash into `pageHash` and `anchorLink`
        2. navigate to `pageHash` (programatically)
        3. scroll to element with id `anchorLink` (if present)

**Comment by vigenere23** [comment]
wow... it actually works! + the default behavior still seems to work... sometimes. It still does create a race condition when using the `restoreScrollState` attribute, but I'm sure we can find a way to prevent this easily (just like the `previousScrollState` check in the `afterUpdate()`?)

<details>
<summary>Here's my prototype for now:</summary>

```js
loc.subscribe(async (newLoc) => {
    // Find if the URL matches a same-page anchor link
    if (newLoc.hash.match(/^#[^\/]\w*/)) {
        if (!lastLoc) {
        // Could be another behavior, but this one seems right to me
            window.location.hash = '#'
            return true // don't know if `true` is needed here
        }

        // Find the "page" (aka fake "location") of last URL
        // Basically removes the anchor link part
        const lastLocPage = lastLoc.hash.split('#')[1]
        // The new URL should be the last page + the added anchor link part
        window.location.hash = lastLocPage + newLoc.hash
        return true // don't know if `true` is needed here
    }

    // Proceed like the rest of code
    lastLoc = newLoc

    // ... rest of function
```

</details>

**Comment by brunnerh** [comment]
Note that jumping to an element via a fragment does not only scroll the element into view, it also changes the tab order context to be that element. So the next time you press tab it will focus the first focusable element within or after the target element.

This is helpful for accessibility features like "skip to content" links.

![image](https://user-images.githubusercontent.com/834235/109417580-8e20d780-79c4-11eb-8270-707e875171c2.png)

I tried reproducing this behavior in JS but so far it turned out to be tricky. If you just try to `focus()` the target element, nothing will happen if the target element itself is not focusable.

**Comment by SeanMcP** [comment]
@brunnerh You'll need to set the tab index of the target element to `0` before calling focus.

**Comment by brunnerh** [comment]
> @brunnerh You'll need to set the tab index of the target element to `0` before calling focus.

Unfortunately that is usually not what you want. Non-interactive elements that are just sections of a page are not supposed to be in the tab order of the document. Maybe one could set the tab index, focus the element and remove the tab index again, but unless that behaves correctly across browsers that is not a viable solution.

**Comment by SeanMcP** [comment]
> @brunnerh one could set the tab index, focus the element and remove the tab index again

Yep! That's a pretty common practice when doing keyboard navigation work in JS.

**Comment by monlasan** [comment]
Good morning/evening ... has a solution been found ? I switched from PageJS to svelte-spa-router because I need hash based routing. Now my documentation anchors links dont work in SSrouter.. Also, If you paste a url in the browser it doesnt work at all

**Comment by MuhammadOwais-SE** [comment]
Is there is a way to remove the "#" in between urls? I want to get clean url.

**Comment by carbogninalberto** [comment]
There is no way to remove the # because it's an hash based router, it exploits the string right after '#' (which is the anchor to an id as standard default html behavior)😁 

If you don't want it, my suggestion is to look into sveltekit (spa mode) or other libraries/frameworks.
